File: DevSupportAndroid/src/main/java/com/sup/dev/android/tools/ToolsAndroid.java
package com.sup.dev.android.tools;

import android.Manifest;
import android.app.ActivityManager;
import android.app.KeyguardManager;
import android.app.UiModeManager;
import android.bluetooth.BluetoothAdapter;
import android.content.ClipData;
import android.content.ClipboardManager;
import android.content.Context;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.content.pm.ResolveInfo;
import android.content.res.Configuration;
import android.content.res.Resources;
import android.media.audiofx.AcousticEchoCanceler;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;
import android.os.Build;
import android.os.Environment;
import android.os.Looper;
import android.os.PowerManager;
import android.support.annotation.RequiresPermission;
import android.view.WindowManager;

import com.sup.dev.android.BuildConfig;
import com.sup.dev.android.app.SupAndroid;
import com.sup.dev.android.magic_box.Miui;
import com.sup.dev.android.magic_box.ServiceNetworkCheck;
import com.sup.dev.java.classes.callbacks.simple.Callback1;
import com.sup.dev.java.libs.debug.Debug;
import com.sup.dev.java.tools.ToolsThreads;

import java.io.BufferedReader;
import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.lang.reflect.Field;
import java.lang.reflect.InvocationTargetException;
import java.util.List;
import java.util.Locale;

import static android.content.Context.WINDOW_SERVICE;

public class ToolsAndroid {

    //
    //  Device
    //

    @RequiresPermission(Manifest.permission.BLUETOOTH)
    public static String getBluetoothMacAddress() {
        BluetoothAdapter bluetoothAdapter = BluetoothAdapter.getDefaultAdapter();
        String bluetoothMacAddress = "";
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.M){
            try {
                Field mServiceField = bluetoothAdapter.getClass().getDeclaredField("mService");
                mServiceField.setAccessible(true);

                Object btManagerService = mServiceField.get(bluetoothAdapter);

                if (btManagerService != null) {
                    bluetoothMacAddress = (String) btManagerService.getClass().getMethod("getAddress").invoke(btManagerService);
                }
            } catch (NoSuchFieldException e) {

            } catch (NoSuchMethodException e) {

            } catch (IllegalAccessException e) {

            } catch (InvocationTargetException e) {

            }
        } else {
            bluetoothMacAddress = bluetoothAdapter.getAddress();
        }
        return bluetoothMacAddress;
    }

    public static String getProcessName() {
        String processName = getProcessNameCmdLine();
        if(processName == null) getProcessNameActivityManager();
        return processName;
    }

    private static String getProcessNameCmdLine() {
        BufferedReader cmdlineReader = null;
        try {
            cmdlineReader = new BufferedReader(new InputStreamReader(new FileInputStream("/proc/" + android.os.Process.myPid() + "/cmdline"), "iso-8859-1"));
            int c;
            StringBuilder processName = new StringBuilder();
            while ((c = cmdlineReader.read()) > 0) processName.append((char) c);
            return processName.toString();
        } catch (Exception ex) {
            Debug.log(ex);
        } finally {
            if (cmdlineReader != null) {
                try {
                    cmdlineReader.close();
                } catch (IOException e) {
                    Debug.log(e);
                }
            }
        }
        return null;
    }

    private static String getProcessNameActivityManager() {
        int pid = android.os.Process.myPid();
        ActivityManager manager = (ActivityManager) SupAndroid.appContext.getSystemService(Context.ACTIVITY_SERVICE);
        List<ActivityManager.RunningAppProcessInfo> infos = manager.getRunningAppProcesses();
        if (infos != null) {
            for (ActivityManager.RunningAppProcessInfo processInfo : infos) {
                if (processInfo.pid == pid) {
                    return processInfo.processName;
                }
            }
        }
        return null;
    }

    @RequiresPermission(android.Manifest.permission.ACCESS_NETWORK_STATE)
    public static String getNetworkName() {
        ConnectivityManager cm = (ConnectivityManager) SupAndroid.appContext.getSystemService(Context.CONNECTIVITY_SERVICE);
        NetworkInfo n = cm.getActiveNetworkInfo();
        return (n == null) ? null : n.getExtraInfo();
    }

    @RequiresPermission(android.Manifest.permission.ACCESS_NETWORK_STATE)
    public static void isHasInternetConnection(Callback1<Boolean> onResult) {
        ServiceNetworkCheck.check(onResult);
    }

    public static int getBottomNavigationBarHeight() {
        Resources resources = SupAndroid.appContext.getResources();
        boolean navBarExists = resources.getBoolean(SupAndroid.appContext.getResources().getIdentifier("config_showNavigationBar", "bool", "android"));
        return navBarExists?resources.getDimensionPixelSize(resources.getIdentifier("navigation_bar_height", "dimen", "android")):0;
    }

    public static boolean isTablet() {
        return (SupAndroid.appContext.getResources().getConfiguration().screenLayout & Configuration.SCREENLAYOUT_SIZE_MASK) >= Configuration.SCREENLAYOUT_SIZE_LARGE;
    }

    public static String getLanguageCode() {
        return Locale.getDefault().getLanguage().toLowerCase();
    }

    public static boolean isEchoCancelerAvailable() {
        return AcousticEchoCanceler.isAvailable();
    }

    public static boolean isDirectToTV() {
        return SupAndroid.appContext.getPackageManager().hasSystemFeature("android.software.leanb‌​ack")
                || SupAndroid.appContext.getPackageManager().hasSystemFeature("android.software.live_tv")
                || ((UiModeManager) SupAndroid.appContext.getSystemService(Context.UI_MODE_SERVICE)).getCurrentModeType() == Configuration.UI_MODE_TYPE_TELEVISION
                || Build.MODEL.toLowerCase().contains("tv box");
    }

    public static boolean appIsVisible() {
        ActivityManager activityManager = (ActivityManager) SupAndroid.appContext.getSystemService(Context.ACTIVITY_SERVICE);
        List<ActivityManager.RunningAppProcessInfo> proc = activityManager.getRunningAppProcesses();

        for (ActivityManager.RunningAppProcessInfo info : proc)
            if (info.processName.equals(SupAndroid.appContext.getPackageName()))
                if (info.importance == ActivityManager.RunningAppProcessInfo.IMPORTANCE_FOREGROUND)
                    return true;

        return false;

    }

    public static boolean isDebug() {
        return BuildConfig.DEBUG;
    }

    public static boolean isMainThread() {
        return Looper.getMainLooper().getThread() == Thread.currentThread();
    }

    public static long getMaxMemory() {
        return Runtime.getRuntime().totalMemory();
    }

    public static long getFreeMemory() {
        Runtime runtime = Runtime.getRuntime();
        return runtime.totalMemory() - runtime.freeMemory();
    }

    public static boolean isMiui() {
        return Miui.isMiui();
    }

    public static void toClipboard(String text) {
        ClipboardManager clipboard = (ClipboardManager) SupAndroid.appContext.getSystemService(Context.CLIPBOARD_SERVICE);
        ClipData clip = ClipData.newPlainText("", text);
        clipboard.setPrimaryClip(clip);
    }

    //
    //  Package / Process
    //

    public static boolean hasBroadcastReceiver(String process, Intent intent, Context context) {
        PackageManager pm = context.getPackageManager();
        List<ResolveInfo> listeners = pm.queryBroadcastReceivers(intent, 0);
        for (ResolveInfo info : listeners) {
            if (info.activityInfo.processName.equals(process))
                return true;
        }
        return false;
    }

    public static boolean checkServiceStarted(String appId, String serviceName) {

        ActivityManager am = (ActivityManager) SupAndroid.appContext.getSystemService(Context.ACTIVITY_SERVICE);

        for (ActivityManager.RunningServiceInfo rsi : am.getRunningServices(150))
            if (appId.equals(rsi.service.getPackageName())
                    && serviceName.toLowerCase().equals(rsi.service.getClassName().toLowerCase()))
                return true;

        return false;
    }

    public static File getSystemRootDir() {
        return Environment.getExternalStorageDirectory();
    }

    //
    //  Screen
    //

    public static boolean isScreenPortrait() {
        return !isScreenLandscape();
    }

    public static boolean isScreenLandscape() {
        return getScreenW() > getScreenH();
    }

    public static int getScreenOrientation() {
        return ((WindowManager) SupAndroid.appContext.getSystemService(WINDOW_SERVICE)).getDefaultDisplay().getRotation();
    }

    public static int getScreenW() {
        return SupAndroid.appContext.getResources().getDisplayMetrics().widthPixels;
    }

    public static int getScreenH() {
        return SupAndroid.appContext.getResources().getDisplayMetrics().heightPixels;
    }

    public static int maxScreenSide() {
        return Math.max(getScreenW(), getScreenH());
    }

    public static int minScreenSide() {
        return Math.min(getScreenW(), getScreenH());
    }

    public static boolean isScreenKeyLocked() {
        KeyguardManager myKeyManager = (KeyguardManager) SupAndroid.appContext.getSystemService(Context.KEYGUARD_SERVICE);
        return myKeyManager.inKeyguardRestrictedInputMode();
    }

    public static void screenOn() {
        if (((PowerManager) SupAndroid.appContext.getSystemService(Context.POWER_SERVICE)).isScreenOn())
            return;

        PowerManager pm = (PowerManager) SupAndroid.appContext.getSystemService(Context.POWER_SERVICE);
        PowerManager.WakeLock screenLock = pm.newWakeLock((PowerManager.SCREEN_BRIGHT_WAKE_LOCK | PowerManager.FULL_WAKE_LOCK | PowerManager.ACQUIRE_CAUSES_WAKEUP), "FULL WAKE LOCK");

        screenLock.acquire();
        new Thread(() -> {
            ToolsThreads.sleep(5000);
            if (screenLock.isHeld())
                screenLock.release();
        }).start();

    }


}

File: DevSupportAndroid/src/main/java/com/sup/dev/android/tools/ToolsAndroid.kt
package com.sup.dev.android.tools

import android.Manifest
import android.app.ActivityManager
import android.app.KeyguardManager
import android.app.UiModeManager
import android.bluetooth.BluetoothAdapter
import android.content.ClipData
import android.content.ClipboardManager
import android.content.Context
import android.content.Context.WINDOW_SERVICE
import android.content.Intent
import android.content.res.Configuration
import android.media.audiofx.AcousticEchoCanceler
import android.net.ConnectivityManager
import android.os.*
import android.support.annotation.RequiresPermission
import android.view.WindowManager
import com.sup.dev.android.BuildConfig
import com.sup.dev.android.app.SupAndroid
import com.sup.dev.android.magic_box.Miui
import com.sup.dev.android.magic_box.ServiceNetworkCheck
import com.sup.dev.java.libs.debug.Debug
import com.sup.dev.java.libs.debug.err
import com.sup.dev.java.tools.ToolsThreads
import java.io.*
import java.lang.reflect.InvocationTargetException
import java.util.*


object ToolsAndroid {

    //
    //  Device
    //

    fun setLanguage(context:Context, lang: String) {
        val res = context.resources
        res.configuration.locale = Locale(lang)
        res.updateConfiguration(res.configuration, res.displayMetrics)
    }

    fun getVersion() = SupAndroid.appContext!!.packageManager.getPackageInfo(SupAndroid.appContext!!.packageName, 0).versionName

    @RequiresPermission(Manifest.permission.BLUETOOTH)
    fun getBluetoothMacAddress(): String {
        val bluetoothAdapter = BluetoothAdapter.getDefaultAdapter()
        var bluetoothMacAddress = ""
        if (android.os.Build.VERSION.SDK_INT >= android.os.Build.VERSION_CODES.M) {
            try {
                val mServiceField = bluetoothAdapter.javaClass.getDeclaredField("mService")
                mServiceField.isAccessible = true
                val btManagerService = mServiceField.get(bluetoothAdapter)
                if (btManagerService != null) bluetoothMacAddress = btManagerService.javaClass.getMethod("getAddress").invoke(btManagerService) as String
            } catch (e: NoSuchFieldException) {
                err(e)
            } catch (e: NoSuchMethodException) {
                err(e)
            } catch (e: IllegalAccessException) {
                err(e)
            } catch (e: InvocationTargetException) {
                err(e)
            }

        } else {
            bluetoothMacAddress = bluetoothAdapter.address
        }
        return bluetoothMacAddress
    }

    fun getProcessName(): String? {
        val processName = getProcessNameCmdLine()
        if (processName == null) getProcessNameActivityManager()
        return processName
    }

    private fun getProcessNameCmdLine(): String? {
        var cmdlineReader: BufferedReader? = null
        try {
            cmdlineReader = BufferedReader(InputStreamReader(FileInputStream("/proc/" + Process.myPid() + "/cmdline"), "iso-8859-1"))
            var c: Int
            val processName = StringBuilder()
            while (true) {
                c = cmdlineReader.read()
                if (c < 1) break
                processName.append(c.toChar())
            }
            return processName.toString()
        } catch (ex: Exception) {
            err(ex)
        } finally {
            if (cmdlineReader != null) {
                try {
                    cmdlineReader.close()
                } catch (e: IOException) {
                    err(e)
                }

            }
        }
        return null
    }

    private fun getProcessNameActivityManager(): String? {
        val pid = android.os.Process.myPid()
        val manager = SupAndroid.appContext!!.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager
        val infos = manager.runningAppProcesses
        if (infos != null) {
            for (processInfo in infos) {
                if (processInfo.pid == pid) {
                    return processInfo.processName
                }
            }
        }
        return null
    }

    @RequiresPermission(android.Manifest.permission.ACCESS_NETWORK_STATE)
    fun getNetworkName(): String? {
        val cm = SupAndroid.appContext!!.getSystemService(Context.CONNECTIVITY_SERVICE) as ConnectivityManager
        val n = cm.activeNetworkInfo
        return n?.extraInfo
    }

    @RequiresPermission(android.Manifest.permission.ACCESS_NETWORK_STATE)
    fun isHasInternetConnection(onResult: (Boolean) -> Unit) {
        ServiceNetworkCheck.check(onResult)
    }

    fun getBottomNavigationBarHeight(): Int {
        val resources = SupAndroid.appContext!!.resources
        val navBarExists = resources.getBoolean(SupAndroid.appContext!!.resources.getIdentifier("config_showNavigationBar", "bool", "android"))
        return if (navBarExists) resources.getDimensionPixelSize(resources.getIdentifier("navigation_bar_height", "dimen", "android")) else 0
    }

    fun isTablet(): Boolean {
        return SupAndroid.appContext!!.resources.configuration.screenLayout and Configuration.SCREENLAYOUT_SIZE_MASK >= Configuration.SCREENLAYOUT_SIZE_LARGE
    }

    fun getLanguageCode(): String {
        return Locale.getDefault().language.toLowerCase()
    }

    fun isEchoCancelerAvailable(): Boolean {
        return AcousticEchoCanceler.isAvailable()
    }

    fun isDirectToTV(): Boolean {
        return (SupAndroid.appContext!!.packageManager.hasSystemFeature("android.software.leanb‌​ack")
                || SupAndroid.appContext!!.packageManager.hasSystemFeature("android.software.live_tv")
                || (SupAndroid.appContext!!.getSystemService(Context.UI_MODE_SERVICE) as UiModeManager).currentModeType == Configuration.UI_MODE_TYPE_TELEVISION
                || Build.MODEL.toLowerCase().contains("tv box"))
    }

    fun appIsVisible(): Boolean {
        val activityManager = SupAndroid.appContext!!.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager
        val proc = activityManager.runningAppProcesses

        for (info in proc)
            if (info.processName == SupAndroid.appContext!!.packageName)
                if (info.importance == ActivityManager.RunningAppProcessInfo.IMPORTANCE_FOREGROUND)
                    return true

        return false

    }

    fun isDebug() = BuildConfig.DEBUG

    fun isMainThread(): Boolean {
        return Looper.getMainLooper().thread === Thread.currentThread()
    }

    fun getMaxMemory(): Long {
        return Runtime.getRuntime().totalMemory()
    }

    fun getFreeMemory(): Long {
        val runtime = Runtime.getRuntime()
        return runtime.totalMemory() - runtime.freeMemory()
    }

    fun isMiui(): Boolean {
        return Miui.isMiui
    }

    fun setToClipboard(text: String) {
        val clipboard = SupAndroid.appContext!!.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
        val clip = ClipData.newPlainText("", text)
        clipboard.primaryClip = clip
    }

    fun getFromClipboard(): String? {
        val clipboard = SupAndroid.appContext!!.getSystemService(Context.CLIPBOARD_SERVICE) as ClipboardManager
        val primaryClip = clipboard.primaryClip
        if (primaryClip == null) return null
        return if (primaryClip.itemCount == 0)
            null
        else {
            val t = primaryClip.getItemAt(0).text
            if(t == null) return null
            return t.toString()
        }
    }


    //
    //  Package / Process
    //

    fun hasBroadcastReceiver(process: String, intent: Intent, context: Context): Boolean {
        val pm = context.packageManager
        val listeners = pm.queryBroadcastReceivers(intent, 0)
        for (info in listeners) {
            if (info.activityInfo.processName == process)
                return true
        }
        return false
    }

    fun checkServiceStarted(appId: String, serviceName: String): Boolean {

        val am = SupAndroid.appContext!!.getSystemService(Context.ACTIVITY_SERVICE) as ActivityManager

        for (rsi in am.getRunningServices(150))
            if (appId == rsi.service.packageName && serviceName.toLowerCase() == rsi.service.className.toLowerCase())
                return true

        return false
    }

    fun getSystemRootDir(): File {
        return Environment.getExternalStorageDirectory()
    }

    //
    //  Screen
    //

    fun isScreenPortrait(): Boolean {
        return !isScreenLandscape()
    }

    fun isScreenLandscape(): Boolean {
        return getScreenW() > getScreenH()
    }

    fun getScreenOrientation(): Int {
        return (SupAndroid.appContext!!.getSystemService(WINDOW_SERVICE) as WindowManager).defaultDisplay.rotation
    }

    fun getScreenW(): Int {
        return SupAndroid.appContext!!.resources.displayMetrics.widthPixels
    }

    fun getScreenH(): Int {
        return SupAndroid.appContext!!.resources.displayMetrics.heightPixels
    }

    fun maxScreenSide(): Int {
        return Math.max(getScreenW(), getScreenH())
    }

    fun minScreenSide(): Int {
        return Math.min(getScreenW(), getScreenH())
    }

    fun isScreenKeyLocked(): Boolean {
        val myKeyManager = SupAndroid.appContext!!.getSystemService(Context.KEYGUARD_SERVICE) as KeyguardManager
        return myKeyManager.inKeyguardRestrictedInputMode()
    }

    fun screenOn() {
        if ((SupAndroid.appContext!!.getSystemService(Context.POWER_SERVICE) as PowerManager).isScreenOn)
            return

        val pm = SupAndroid.appContext!!.getSystemService(Context.POWER_SERVICE) as PowerManager
        val screenLock = pm.newWakeLock(PowerManager.SCREEN_BRIGHT_WAKE_LOCK or PowerManager.FULL_WAKE_LOCK or PowerManager.ACQUIRE_CAUSES_WAKEUP, "FULL WAKE LOCK")

        screenLock.acquire()
        Thread {
            ToolsThreads.sleep(5000)
            if (screenLock.isHeld)
                screenLock.release()
        }.start()

    }


}